Playwright Initial Setup

panoramix360 opened the pull request:
This PR introduces some initial configurations to use this project as a utility for other Plugins in the Mattermost Space.

It adds:
- Testcontainers integration
- Playwright initial configuration
- MattermostContainer that is used by the Plugins to initialize Mattermost to use in E2E tests

As an initial configuration this is already being used by me on another branch inside the `mattermost-plugin-todo` to create an initial test.

panoramix360 committed:
feat: first playwright config

panoramix360 committed:
feat: adding utils files

panoramix360 committed:
feat: change the export way

panoramix360 committed:
adds config to the RunContainer function

panoramix360 committed:
feat: configuring the node-fetch package

mickmister commented on src/index.ts:
```diff
@@ -0,0 +1,6 @@
+import fetch from 'node-fetch'
+global.fetch = fetch
+
+export {default as MattermostContainer} from './mmcontainer'
+export {default as RunContainer} from './plugincontainer'
+export * from './utils'
```
Exports from the package :+1:

panoramix360 replied:
Do you want me to change anything about exports?

mickmister replied:
Nope I'm just outlining it with this PR comment. Just to surface what's exported into the "Conversation" tab of the PR

mickmister commented on src/plugincontainer.ts:
```diff
@@ -0,0 +1,74 @@
+  if (filename === "") {
+      throw("No tar.gz file found in dist folder")
+  }
+  const pluginConfig = {
+    "clientid":                   "client-id",
+  	"clientsecret":               "client-secret",
+  	"connectedusersallowed":      1000,
+  	"encryptionkey":              "eyPBz0mBhwfGGwce9hp4TWaYzgY7MdIB",
+  	"maxSizeForCompleteDownload": 20,
+  	"maxsizeforcompletedownload": 20,
+  	"tenantid":                   "tenant-id",
+  	"webhooksecret":              "webhook-secret",
+  }
```
This config is specific to the msteams plugin right? To make this be a more realistic test we can see in action, maybe we can put an `example_plugin_tests` directory in this repo. This removes the hurdle of involving another plugin repo to see what the developer experience is like when we integrate with a plugin project.

We can use `Client4.installPluginFromUrl` if we don't want to have an actual plugin project in this repo. We can use the following url from this demo plugin release https://github.com/mattermost/mattermost-plugin-demo/releases/tag/v0.10.0:

`Client4.installPluginFromUrl('https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.10.0/com.mattermost.demo-plugin-0.10.0.tar.gz')`

Then we can write playwright tests in this repo, as if they were defined for the demo plugin in the demo plugin repo. What do you think?

panoramix360 replied:
That's a good idea!

I agree, I'm making the `pluginConfig` part of the configuration that you need to pass on the RunContainer as well, what do you think?

panoramix360 replied:
I'm already testing that, but I'm receiving this error:
![image](https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/assets/2154092/f2df1816-cda8-4284-8714-ec43a551fe12)

Do you know what it might be?

mickmister replied:
It seems the install url may be the issue. We should use https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.10.0/com.mattermost.demo-plugin-0.10.0.tar.gz instead of https://github.com/mattermost/mattermost-plugin-demo/releases/tag/v0.10.0. I was a bit confusing with my suggestion above, as I mentioned two different URLs

mickmister replied:
> I'm making the `pluginConfig` part of the configuration that you need to pass on the RunContainer as well, what do you think?

I think we would want to make the most generic form of configuration not assume this container is for plugin testing, so that could come with something like a `withPluginConfig` method

panoramix360 replied:
Understood, probably I'll need to change a lot of the structure as it is right now then

Let me think about this some more then.

panoramix360 replied:
Can you take a look at the new structure? I think that it fixes this.

mickmister commented on playwright.config.ts:
```diff
@@ -0,0 +1,77 @@
+/**
+ * See https://playwright.dev/docs/test-configuration.
+ */
+export default defineConfig({
+  testDir: './tests',
+  /* Run tests in files in parallel */
+  fullyParallel: true,
+  /* Fail the build on CI if you accidentally left test.only in the source code. */
+  forbidOnly: !!process.env.CI,
+  /* Retry on CI only */
```
If we go the route of having the `example_plugin_tests` I suggested, we would move this file into that folder, since I think the actual playwright config would still exist in the individual plugin projects. We can define things we want to share in a file like this though, if there are things specific to plugin projects that we want to centralize in the playwright config.

I think general playwright config makes sense to duplicate among the plugin projects, partially because it's pretty minimal, but also because that gives plugin projects the freedom to configure playwright itself the way it wants to. imo the main thing about this repo is to provide utilities that are useful for plugin projects, such as utility functions and Mattermost test environment bootstrapping that we have implemented in this PR

panoramix360 replied:
Understood, I'll create this file inside the example tests, this doesn't need to be here.

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    stop = async () => {
+        await this.pgContainer.stop()
```
I was going to say we need to clean up the database connection that is created in the `db()` function somewhere (possibly in the test itself), but stopping the container may be a fine alternative to that

mickmister replied:
Do the containers get cleaned up if the test crashes? I don't know enough about `testcontainers` to know that

panoramix360 replied:
I'm going to investigate that.

But if I could guess, probably they are

panoramix360 replied:
@mickmister 
I did some digging and it seems the `stop()` function is from the `AbstractStartedContainer` class from the Testcontainers library.

This function just stops the container.

Since the developer will run the test again if it crashes, everything will be created again, because the `start()` function inside the MattermostContainer already starts everything again, creating every container.

So I don't know if we need to worry about that for now at least.

But probably in the future we could enable more features and things for the developer so he can control the database if he wants.

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+        await this.network.stop()
+    }
+
+    createAdmin = async (email: string, username: string, password: string) => {
+        await this.container.exec(["mmctl", "--local", "user", "create", "--email", email, "--username", username, "--password", password, "--system-admin", "--email-verified"])
+    }
```
Can we abstract further in a separate method called `createDefaultAdminUser`? So the plugin project doesn't need to make any decisions about any credentials

panoramix360 replied:
It's already addressed by the code in a different way.

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+        await this.container.exec(["mmctl", "--local", "config", "set", "ServiceSettings.ListenAddress", `${containerPort}`])
+    }
+
+    installPlugin = async (pluginPath: string, pluginID: string, pluginConfig: any) => {
```
We could have two methods `installPluginFromLocalBinary` and `installPluginFromUrl` if we want to have two different ways to install the plugin

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+        return this
+    }
+
+    constructor() {
```
Probably good to have the `constructor` be the first function defined in the class. Just my opinion though. If you have reasons otherwise I understand

panoramix360 replied:
I agree 100%, done!

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+                "MM_LOGSETTINGS_CONSOLELEVEL":        "DEBUG",
+                "MM_LOGSETTINGS_FILELEVEL":           "DEBUG",
+        };
+        this.email = defaultEmail;
+        this.username = defaultUsername;
+        this.password = defaultPassword;
+        this.teamName = defaultTeamName;
+        this.teamDisplayName = defaultTeamDisplayName;
```
Nvm about the `createDefaultAdminUser`. It should be obvious how to "get" the default admin after calling `start` though. `getAdminClient` may be enough. The developer could always call `adminClient.getMe()` but it may make sense to store that object somewhere in advance, as you may want to know user's generated id for instance. May be premature optimization and clutter the classes namespace though

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+            .withNetworkAliases("db")
+            .start()
+
+        this.container = await new GenericContainer(defaultMattermostImage)
```
Would be good to be able to inject a value for the MM image to use

panoramix360 replied:
What do you think of adding this feature as a separate issue?

it seems interesting, but I think this PR is getting big

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+            .withNetworkAliases("db")
+            .start()
+
+        this.container = await new GenericContainer(defaultMattermostImage)
+            .withEnvironment(this.envs)
+            .withExposedPorts(8065)
+            .withNetwork(this.network)
+            .withNetworkAliases("mattermost")
+            .withCommand(this.command)
+            .withWaitStrategy(Wait.forLogMessage("Server is listening on"))
+            .withCopyFilesToContainer(this.configFile)
+            .start()
```
I really like this chained thing from constructor, to `with...()` methods, to `start()` all in one operation :+1:

panoramix360 replied:
Yep, that's really nice, I'll try to adopt this on other places as well when needed.

mickmister commented on src/mmcontainer.ts:
```diff
@@ -0,0 +1,189 @@
+            await this.installPlugin(plugin.path, plugin.id, plugin.config)
+        }
+
+        return this
```
Probably want to introduce something like eslint (that's probably the best choice since that what we use in other projects) to make sure we're consistent with coding styles, e.g. using semicolons or not

mickmister commented on src/plugincontainer.ts:
```diff
@@ -0,0 +1,74 @@
+
+const RunContainer = async ({ packageName, distPath }: RunContainerConfig): Promise<MattermostContainer> => {
+  let filename = "";
+  fs.readdirSync(distPath).forEach(file => {
+      if (file.endsWith(".tar.gz")) {
+          filename = distPath + file
+      }
+  })
```
Probably good to use a `for const` loop and break when we find when we're looking for, mainly for readability

panoramix360 replied:
done!

mickmister commented on src/plugincontainer.ts:
```diff
@@ -0,0 +1,74 @@
+  distPath: string
+}
+
+const RunContainer = async ({ packageName, distPath }: RunContainerConfig): Promise<MattermostContainer> => {
```
I'm thinking we may want to be able to inject more values upon calling this function, but I'm not sure which ones at the moment. This could be exactly what we land on

panoramix360 replied:
I already added some of that, can you take a look at the code again?

But probably I'm going to change that slightly, I'll explain in the comment below on the PR.

mickmister commented on src/plugincontainer.ts:
```diff
@@ -0,0 +1,74 @@
+  await adminClient.savePreferences(admin.id,[
+    {user_id: admin.id, category: 'tutorial_step', name: admin.id, value: '999'},
+    {user_id: admin.id, category: 'onboarding_task_list', name: 'onboarding_task_list_show', value: 'false'},
+    {user_id: admin.id, category: 'onboarding_task_list', name: 'onboarding_task_list_open', value: 'false'},
+    {
+        user_id: admin.id,
+        category: 'drafts',
+        name: 'drafts_tour_tip_showed',
+        value: JSON.stringify({drafts_tour_tip_showed: true}),
+    },
+    {user_id: admin.id, category: 'crt_thread_pane_step', name: admin.id, value: '999'},
```
Probably want to have this sort of thing be consistent too. Single-line vs multi-line

panoramix360 replied:
I think when we add the ESLint and Prettier as mentioned on the comment below, this will get fixed, thanks!

mickmister reviewed (commented):
Awesome work Lucas! 🚀 

I have some comments for discussion, including one about putting some example tests in this repo to see this in action! Can't wait to use this awesome functionality 😄

panoramix360 committed:
feat: externalizing PluginConfig

panoramix360 committed:
feat: adding external plugin configuration

panoramix360 committed:
feat: adding an example-test-folder

panoramix360 committed:
chore: removes the console log

panoramix360 committed:
feat: moving the constructor to the top and changing the name of the example tests

panoramix360 commented:
Thanks for your comments @mickmister 

In the last few commits, I externalized the PluginConfig, added the external plugin configuration, and added the example-test folder to have tests in it.

After reading all of your comments, I think my plan of action will be:
- Analyze thoughtfully the code in `mmcontainer.ts` and remove everything that it's plugin specific related to make it more generic
- Create and reorganize the `plugincontainer.ts` to be plugin agnostic
- Check to see if the containers are cleaned after a test failure
- Address the comments to change code in specific places
- Add ESLint to the project and maybe even Prettier to code format (@mickmister can I borrow that from another project? If so, can you point me in the right direction? I can probably search for it as well.)

I'm thinking that maybe the `mmcontainer.ts` could be called `mattermost_container.ts` and this will be responsible for:
- Having code that interacts with the `testcontainers` library
- Have generic code that interacts with the mattermost container in ways that it makes sense for Mattermost functions (like creating new users, creating new channels etc)
- Be plugin-agnostic so we can use that on other tests as well, and not only on plugins.

And the `plugincontainer.ts` can be renamed to `run_plugin_container.ts` and this will be responsible for:
- Have a method called `RunPluginContainer`code that leverages a Mattermost container using the `mattermost_container.ts` to create a Mattermost instance with a plugin/plugins already installed
- Have utilities that work for plugins and let the dev being able to test the plugin in this way
- Be plugin-agnostic but have a customizable way to interact with the plugin

Is this a good plan?

What do you think?

cc: @mickmister @jespino

Thank you!

mickmister commented:
@panoramix360 For eslint, this file from the GitHub plugin is probably good minus the React stuff https://github.com/mattermost/mattermost-plugin-github/blob/master/webapp/.eslintrc.json. I personally don't recommend adding prettier, mainly because eslint has been made standard in our projects with no use of prettier. eslint does everything we have deemed that we need in terms of formatting, even if eslint's primary use case is not for formatting.

> I'm thinking that maybe the `mmcontainer.ts` could be called `mattermost_container.ts` and this will be responsible for:
> - Having code that interacts with the `testcontainers` library
> - Have generic code that interacts with the mattermost container in ways that it makes sense for Mattermost functions (like creating new users, creating new channels etc)
> - Be plugin-agnostic so we can use that on other tests as well, and not only on plugins.

This all sounds good :+1:

> And the `plugincontainer.ts` can be renamed to `run_plugin_container.ts` and this will be responsible for:
> - Have a method called `RunPluginContainer`code that leverages a Mattermost container using the `mattermost_container.ts` to create a Mattermost instance with a plugin/plugins already installed
> - Have utilities that work for plugins and let the dev being able to test the plugin in this way
> - Be plugin-agnostic but have a customizable way to interact with the plugin

For the terminology here, this makes me think that we are running the plugin in a separate docker container, which is obv not what's happening. Would the developer still have access to the main `MattermostContainer`? Would the plugin one subclass this, or simply have a reference to it? Maybe we can have a `withPlugin` method or similar on `MattermostContainer` that somehow returns or provides the reference to the plugin helper object. Or maybe that plugin class can accept a `MattermostContainer` as an argument to its constructor.

panoramix360 commented:
Thanks for confirming it, I think it's easier to work with a path clearer forward 😄 

> @panoramix360 For eslint, this file from the GitHub plugin is probably good minus the React stuff https://github.com/mattermost/mattermost-plugin-github/blob/master/webapp/.eslintrc.json. I personally don't recommend adding prettier, mainly because eslint has been made standard in our projects with no use of prettier. eslint does everything we have deemed that we need in terms of formatting, even if eslint's primary use case is not for formatting.

Ok, no Prettier then! I noticed that when I worked on the `mattermost-mobile` project. But let me ask you one question, using just ESLint, it's possible to configure VS Code to format automatically on save? or do we need to run the `fix` command?

>For the terminology here, this makes me think that we are running the plugin in a separate docker container, which is obv not what's happening. Would the developer still have access to the main MattermostContainer? Would the plugin one subclass this, or simply have a reference to it? Maybe we can have a withPlugin method or similar on MattermostContainer that somehow returns or provides the reference to the plugin helper object. Or maybe that plugin class can accept a MattermostContainer as an argument to its constructor.

I agree with that approach, I'll align my code to that. I just was going to use mostly of the structure the code that was already in place regarding the `RunContainer` since this was the way it was done before.

But I agree that maybe it's better for the dev to interact just with the MattermostContainer itself and get an instance of the plugin and interact with it how it sees fit.

This week I'm trying to organize the code and follow the approaches that we discussed, thanks for the insights!

panoramix360 committed:
feat: enhancing the readfile for loop

panoramix360 committed:
feat: configuring eslint and formatting

panoramix360 committed:
chore: removes unecessary imports

panoramix360 committed:
feat: removing the `no-process-env` and `no-undefined` to not complain about the `playwright.config.ts` plus fixing some lint errors

panoramix360 commented:
Added the ESLInt configuration!

I'll continue with the other implementations.

panoramix360 committed:
feat: creating the concept of the MattermostPlugin

panoramix360 committed:
fix: eslint problem on the promises of the plugins to install

panoramix360 committed:
refactor to make it simpler to use an external plugin

panoramix360 commented:
@mickmister

done! can you take a look at the new structure?

I think it solves a lot of scaling problems within the code, and separates well the MattermostPlugin vs MattermostContainer

I liked the new structure, probably we will have some things to still adjust and enhance regarding other topics, but as a base, I think this is beginning to be more solid.

The tests are passing already. What more tests do you think we could create inside the demo-plugin?

Please share your thoughts so I can improve this if needed.

mickmister commented on package.json:
```diff
@@ -6,10 +6,21 @@
+    "babel-eslint": "^10.1.0",
+    "eslint": "^8.57.0",
+    "eslint-plugin-import": "^2.29.1",
+    "eslint-plugin-n": "^16.6.2",
```
Wonder what `eslint-plugin-n` is?

panoramix360 replied:
It's additional eslint rules for Node.JS but since we are not using it on the .eslintrc I removed.

Thanks, probably it was added by the default eslint configuration.

mickmister commented on src/index.ts:
```diff
@@ -1,6 +1,6 @@
\ No newline at end of file
+export {default as MattermostContainer} from './mattermostcontainer';
+export * from './plugin';
+export * from './utils';
```
Nit: No newline at the end of the file. Maybe `eslint` can be configured to find this? This is in other files as well

Btw you can set this in VSCode automatically with these settings:

```json
    "files.insertFinalNewline": true,
    "files.trimFinalNewlines": true,
```

panoramix360 replied:
[It seems this rule is deprecated on ESLint, strange.](https://eslint.org/docs/latest/rules/eol-last)

To add to eslint we would need to add the [@stylistic/eslint-plugin-ts](https://eslint.style/packages/js)

Do you think it's doable?

Already changed VS Code to add new lines.

mickmister commented on src/index.ts:
```diff
@@ -1,6 +1,6 @@
-import fetch from 'node-fetch'
-global.fetch = fetch
+import fetch from 'node-fetch';
+global.fetch = fetch;
```
Maybe we can do something like this to make sure we only do this when we need to. Some versions of node already have this, and if the developer wants to use that version of `fetch`, then we shouldn't overwrite that version

```suggestion
import fetch from 'node-fetch';

if (typeof fetch === 'undefined') {
    global.fetch = fetch;
}
```

panoramix360 replied:
ESLint is complaining:
<img width="1198" alt="image" src="https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/assets/2154092/112622f3-94da-4f88-8e6a-0130d0873d26">

Is this code correct? 😄 

If `fetch` is undefined, probably we don't want to use it to replace global?

Since the fetch reference here is from the import.

panoramix360 replied:
@mickmister any thoughts?

mickmister replied:
Maybe try this:

```ts
if (typeof fetch === 'undefined') {
    (global as any).fetch = fetch;
}
```

mickmister commented on src/plugin/mattermostplugin.ts:
```diff
@@ -0,0 +1,46 @@
+export default class MattermostPlugin {
+    config: Config;
+    isExternal: boolean = false;
+    path: string;
```
Typecheck CI should catch that `this.path` is not set in the constructor. Probably should add typechecking to CI. Do we need `this.path` anyway?

panoramix360 replied:
Yep, we need it.

The idea behind this path is that it will be updated accordingly with what the user wants to use.

The different options are external path and local binary, both of these have different logic to generate the path:
```
withExternalPath = (externalPath: string): MattermostPlugin => {
        this.isExternal = true;
        this.path = externalPath;

        return this;
    };

    withLocalBinary = (path: string): MattermostPlugin => {
        let filename = '';
        const files = fs.readdirSync(path);
        for (const file of files) {
            if (file.endsWith('.tar.gz')) {
                filename = path + file;
                break;
            }
        }
        if (filename === '') {
            throw Error('No tar.gz file found in dist folder');
        }

        return this;
    };
```

Makes sense?

mickmister replied:
I guess I don't see where `this.path` is used anywhere though. Is it used externally from the class?

panoramix360 replied:
It's used on the `mattermostcontainer.ts` t o install the plugins:

```
installPluginFromLocalBinary = async (plugin: MattermostPlugin<unknown>) => {
        const {pluginId} = plugin.options;

        const patch = JSON.stringify({PluginSettings: {Plugins: {[pluginId]: plugin.options.pluginConfig}}});

        await this.container.copyFilesToContainer([{source: plugin.path, target: '/tmp/plugin.tar.gz'}]);
        await this.container.copyContentToContainer([{content: patch, target: '/tmp/plugin.config.json'}]);

        await this.container.exec(['mmctl', '--local', 'plugin', 'add', '/tmp/plugin.tar.gz']);
        await this.container.exec(['mmctl', '--local', 'config', 'patch', '/tmp/plugin.config.json']);
        await this.container.exec(['mmctl', '--local', 'plugin', 'enable', pluginId]);
    };

    installPluginFromUrl = async (plugin: MattermostPlugin<unknown>) => {
        const client = await this.getAdminClient();
        const manifest = await client.installPluginFromUrl(plugin.path);
        await this.container.exec(['mmctl', '--local', 'plugin', 'enable', manifest.id]);
    };
```

mickmister replied:
Do you think we can set the property to `public` to signal that it's meant to be used by another context that has a reference to this object?

panoramix360 replied:
done!

mickmister commented on src/plugin/mattermostplugin.ts:
```diff
@@ -0,0 +1,46 @@
+        this.config = config;
+    }
+
+    withExternalPath = (externalPath: string): MattermostPlugin => {
```
I would call this `withExternalURL`

panoramix360 replied:
done!

mickmister commented on example_plugin_tests/tests/example.spec.ts:
```diff
@@ -1,39 +1,42 @@
-    }
-  });
+    demoPluginInstance = new MattermostPlugin({
+        packageName: 'com.mattermost.demo-plugin',
```
```suggestion
        pluginId: 'com.mattermost.demo-plugin',
```

panoramix360 replied:
done!

mickmister commented on example_plugin_tests/tests/example.spec.ts:
```diff
@@ -1,39 +1,42 @@
+        maxSizeForCompleteDownload: 20,
+        maxsizeforcompletedownload: 20,
+        tenantid: 'tenant-id',
+        webhooksecret: 'webhook-secret',
```
These config values are unrelated to the framework here. As an aside, I think we need to support providing a plugin config. We could do something like:

```ts
type DemoPluginConfig = {
    // ...
}

test.beforeAll(async () => {
    demoPluginInstance = new MattermostPlugin<DemoPluginConfig>({
        packageName: 'com.mattermost.demo-plugin',
        pluginConfig: {
            // ...implements DemoPluginConfig type
        },
        // ...
    });
```

Then `MattermostPlugin` can use that "generic" to make sure `pluginConfig` matches that type

```ts
type MattermostPluginOptions<PluginConfigType> = {
    pluginId: string;
    pluginConfig: PluginConfigType;
}

class MattermostPlugin<PluginConfigType> {
    constructor(options: MattermostPluginOptions<PluginConfigType>) {
        // ...
    }
}
```

panoramix360 replied:
That's an awesome idea, thank you!

Just changed the code, take a look :)

And I even learned what the `unknown` in TypeScript is for hehe loved it

mickmister commented on example_plugin_tests/tests/example.spec.ts:
```diff
@@ -1,39 +1,42 @@
+        tenantid: 'tenant-id',
+        webhooksecret: 'webhook-secret',
+    }).
+        withExternalPath('https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.10.0/com.mattermost.demo-plugin-0.10.0.tar.gz');
```
Probably good to put this as a constant at the top of this file

panoramix360 replied:
done!

mickmister commented on example_plugin_tests/tests/example.spec.ts:
```diff
@@ -1,39 +1,42 @@
+    }).
+        withExternalPath('https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.10.0/com.mattermost.demo-plugin-0.10.0.tar.gz');
+
+    mattermost = await new MattermostContainer().
+        withPlugin(demoPluginInstance).
+        startWithUserSetup();
```
Excellent DX 😄

panoramix360 replied:
Thank you Mike 😃

mickmister commented on src/mattermostcontainer.ts:
```diff
@@ -0,0 +1,246 @@
+    teamDisplayName: string;
+    envs: {[key: string]: string};
+    command: string[];
+    configFile: any[];
```
`any[]` seems a bit weird here. I'm thinking maybe we want to provide `config` instead, which is available through `@mattermost/types`. Then the plugin itself can take care of reading a file if it needs to. Not sure though. @jespino thoughts on this?

panoramix360 replied:
Yeah, I leave it like that because I didn't get the full picture of what this is doing actually.

Can we create a different issue to discuss it?

Probably we would want to do more customizations for the developer to use.

mickmister replied:
Created an issue for this https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/issues/4

mickmister commented on src/mattermostcontainer.ts:
```diff
@@ -0,0 +1,246 @@
+            MM_PASSWORDSETTINGS_MINIMUMLENGTH: '5',
+            MM_PLUGINSETTINGS_ENABLEUPLOADS: 'true',
+            MM_FILESETTINGS_MAXFILESIZE: '256000000',
+            MM_LOGSETTINGS_CONSOLELEVEL: 'DEBUG',
+            MM_LOGSETTINGS_FILELEVEL: 'DEBUG',
```
Are we able to access the server logs when debugging e2e tests? Also, is it possible to instead of using the MM test container, and instead run MM locally? This is so the dev can edit the server code during writing/debugging a test

panoramix360 replied:
That's a good idea, probably we could work on that.

What do you think of creating a separate issue to tackle that so we don't make this PR too big?

mickmister replied:
Ticket here https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/issues/7

panoramix360 replied:
I am solving this so we can discuss the issue on the ticket

mickmister commented:
> What more tests do you think we could create inside the demo-plugin?

@DHaussermann Can you please provide some guidance on where we can find the documented test cases for the demo plugin? @panoramix360 We can pick 1 or 2 of the simplest cases to implement here

> it's possible to configure VS Code to format automatically on save? or do we need to run the fix command?

Yep it's a configurable setting for the eslint plugin https://www.digitalocean.com/community/tutorials/workflow-auto-eslinting

I typically type `Cmd Shift P`, Type `eslint`, `Enter` because I don't like the editor changing my stuff while I'm still typing haha. I hit `Cmd S` quite often 🙂

mickmister reviewed (commented):
Hi @panoramix360, this is looking great! 🚀 
I have a few more comments on the PR. Let me know what you think 😄

Thanks so much @panoramix360!

mickmister commented:
> What more tests do you think we could create inside the demo-plugin?

@panoramix360 Maybe we can run one of the `/dialog` commands (even just `/dialog` which shows a dialog with all of the dialog element types), fill out the form, and submit. And then verify the created ephemeral message. This is also good so we can share some utility functions in the library like filling out interactive dialogs

mickmister commented:
FYI there is some existing code in the GitHub plugin's e2e tests that works with interactive dialogs that we can share here

panoramix360 committed:
chore: fix babel old version

panoramix360 committed:
chore: setting the nvm version

panoramix360 committed:
chore: actions on save

panoramix360 committed:
chore: removing eslint n

panoramix360 committed:
chore: fixing babel es lint parser

panoramix360 committed:
newline in the end of file

panoramix360 committed:
chore: renaming

panoramix360 committed:
chore: rename pluginid

panoramix360 committed:
feat: changing how the client types the configs for the plugin

panoramix360 committed:
feat: adding demo plugin as a constant

mickmister commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+20.11
```
A little concerning to me that this is a lot further along than most plugin projects (typically 16.x IIRC). Though e2e test code should be fine to be different than the npm versions used to compile the plugins

panoramix360 replied:
Ok, I could lower this version down to 16.0 then. I'll test it.

panoramix360 replied:
just checked, but it seems that the `testcontainers-node` needs v20 higher:
https://github.com/testcontainers/testcontainers-node/blob/main/.nvmrc

Tried using the v16, and the container didn't work.

![image](https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/assets/2154092/590009f1-2973-46a8-9979-71ac251d5894)

any ideas? I can try testing it with v18.

panoramix360 replied:
Ignore my last comment haha

It worked with v16.20, just committed the nvmrc

I was not running docker 🐳

mickmister replied:
If we theoretically needed to change it for e2e tests, we could add a `.nvmrc` file specifically in the e2e test dir

mickmister commented on .vscode/settings.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": "always"
```
This is a little opinionated and not something that I would want personally. I like typing `Cmd Shift P` and then typing `eslint` to fix the issues. I think this setting should be set on developers' machines separately. But maybe the user settings override this and it's not something I need to worry about?

panoramix360 replied:
Workspace settings always replace the user settings.

I think you are right, about removing it.

Just commited.

mickmister reviewed (commented):
This looking awesome @panoramix360!!

I'll leave to you to decide if we need to implement the demo plugin tests. Maybe we can at least verify that the `/dialog` slash command exists in the command autocomplete to show that the demo plugin was correctly installed.

One thing I want to be sure of before merging (or maybe after this PR) is that this is usable from external repos. Do you mind trying to import this code from a repo like demo plugin or GitHub plugin e2e tests? You and I agreed that we can keep this as Typescript and do not need to compile it, but maybe we should via a `postinstall` script in this repo. I'd like to importing with its current state first and see if we run into any issues. We would install the npm dependency as a git-based dependency right now since we don't have this published on npm.

Great work Lucas!!

panoramix360 committed:
chore: changing the default node version and the settings

panoramix360 commented:
hey @mickmister! thanks for the feedback!

I worked on some minor details and now it's just missing the `/dialog` tests, I'll try to do it in these next few days so we can merge this PR.

Thank you!

mickmister commented on src/mattermostcontainer.ts:
```diff
@@ -0,0 +1,246 @@
+    installPluginFromUrl = async (plugin: MattermostPlugin<unknown>) => {
+        const client = await this.getAdminClient();
+        const manifest = await client.installPluginFromUrl(plugin.path);
+        await this.container.exec(['mmctl', '--local', 'plugin', 'enable', manifest.id]);
```
I think we'll want to add a `--force` flag here, to ensure the server downloads the plugin, even if it's already installed. We want this, because otherwise the server won't update the plugin if given one with the same id. Same with the `plugin add` command above

panoramix360 replied:
Makes sense! added!

panoramix360 committed:
feat: increasing the timeout and creating a dialog test

panoramix360 committed:
feat: adding the env

panoramix360 committed:
feat: adding a force boolean on the install commands

panoramix360 commented:
@mickmister 

I added a test example after some effort, I ran into some problems but already solved talking through the community, thank you for that

You can already review the latest changes.

I'll work on running this on the CI now, let's see if it's easy enough.

panoramix360 committed:
feat: testing github actions

panoramix360 committed:
feat: change workflow action

panoramix360 committed:
ci: moving the playwright workflow

panoramix360 committed:
ci: test another command

panoramix360 committed:
ci: adding a working dir

panoramix360 committed:
ci: playwright report path

panoramix360 commented:
hey @mickmister 

The CI is configured and passing already, really happy with it.

It's taking shape! 

Do you think it's good to merge?

mickmister commented on .github/workflows/playwright.yml:
```diff
@@ -3,7 +3,8 @@ on:
     branches: [ main, master ]
   pull_request:
-    branches: [ main, master ]
+  workflow_dispatch:
```
I wonder if it makes sense to have this workflow be in a centralized GH action for other plugin projects to use. The one thing is that the way we set up our shared actions, the caller cannot pass any arbitrary environment variables. I think that works fine since pretty much all configuration is done in the actual test now due to using `testcontainers` :tada: :tada:

Made a ticket for creating a centralized action/workflow to run playwright tests https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/issues/6

panoramix360 replied:
That makes sense! let's work on that in the new ticket!

mickmister replied:
Keeping this conversation unresolved to surface the discussion

mickmister commented on .github/workflows/playwright.yml:
```diff
@@ -13,15 +14,20 @@ jobs:
     - uses: actions/setup-node@v4
       with:
         node-version: lts/*
```
We should use the `.nvmrc` version like so https://github.com/mattermost-community/mattermost-plugin-todo/blob/b8dff94f7bebfcb7f81f36f588f110e08e34ef9f/.github/workflows/playwright.yml#L111

panoramix360 replied:
yep, done!

mickmister replied:
The workflow should ideally use the nvmrc file directly like here https://github.com/mattermost-community/mattermost-plugin-todo/blob/b8dff94f7bebfcb7f81f36f588f110e08e34ef9f/.github/workflows/playwright.yml#L111

It begs the question, should the playwright tests necessarily be run with the same node version as the `.nvmrc` file? We probably would want a separate `.nvmrc` file in the playwright directory so there is no coupling there. The node version that the plugin is compiled with, and the node version that will run the playwright tests are pretty unrelated.

Not sure what to choose for "which node version should run the playwright tests". @hanzei @hmhealey Any thoughts on this?

mickmister commented on .github/workflows/playwright.yml:
```diff
@@ -13,15 +14,20 @@ jobs:
+      working-directory: example_plugin_tests/
     - name: Run Playwright tests
-      run: npx playwright test
+      run: npm run test
```
It would be nice to have some linting and type-checking CI processes on this repo as well. Can be done in a separate PR if you think that makes sense

panoramix360 replied:
can we do it on another issue? I think this PR is getting big hehe

Created the issue: https://github.com/mattermosttest/mattermost-plugin-e2e-test-utils/issues/8

mickmister commented on example_plugin_tests/playwright.config.ts:
```diff
@@ -27,6 +27,9 @@ export default defineConfig({
     reporter: 'html',
 
+    /* More timeout since raising containers are costly  */
+    timeout: 50000,
```
What exactly is this timeout scoped to? The overall running of tests? Maybe we can set a timeout on the setup specifically somehow?

mickmister replied:
I think we may want rename this folder to `demo_plugin_tests`

panoramix360 replied:
Yep, the timeout is to run the tests, since they take a little time to create the containers, etc, we need this extra time for the tests to work. I was running into some timed-out tests because of being only 30 seconds.

We could potentially add this as a setup configuration like you said.

I have renamed the folder as well!

mickmister reviewed (approved):
I think this is good to go. Let's ask @jespino for another review then I think we can merge 👍 🚀

panoramix360 committed:
chore: public var

panoramix360 committed:
feat: changing node version to equal to nvmrc

panoramix360 committed:
chore: renaming folder

panoramix360 committed:
feat: adding the dist files

panoramix360 committed:
feat: updates package json

panoramix360 committed:
fix: issue when it's a local binary, the path was not being filled

panoramix360 committed:
feat: update MattermostContainer to set user ID after login and clean up console logs
